src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -27,22 +27,37 @@ object JoinIndexRanker {
    * Rearranges all available index options according to their cost. The first one is the best
    * with minimum cost.
    *
-   * Pick the one with least amount of shuffling and most number of buckets. When two indices have
-   * same number of buckets, there is zero shuffling. If number of buckets differ, one of the
-   * indices gets reshuffled into the number of buckets equal to the other. Secondly, more the
-   * number of buckets, better the parallelism achieved during join in general, assuming there is
-   * no resource constraint.
+   * If hybridScanEnabled is true, check the number of source files for each index. If the number
+   * of source files are different, pick the index pair which covers the most source data so that
+   * we could minimize the amount of data for on-the-fly shuffle or merge.
+   * Otherwise, pick the one with least amount of shuffling and most number of buckets. When two
+   * indices have same number of buckets, there is zero shuffling. If number of buckets differ,
+   * one of the indices gets reshuffled into the number of buckets equal to the other. Secondly,
+   * more the number of buckets, better the parallelism achieved during join in general, assuming
+   * there is no resource constraint.
+   *
    *
    * @param indexPairs index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
+   * @param hybridScanEnabled HybridScan config.
    * @return rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
-      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
+      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)],
+      hybridScanEnabled: Boolean): Seq[(IndexLogEntry, IndexLogEntry)] = {
 
     indexPairs.sortWith {
       case ((left1, left2), (right1, right2)) =>
-        if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
+        lazy val numSourceFilesLeft =
+          left1.allSourceFileInfos.size + left2.allSourceFileInfos.size
+        lazy val numSourceFilesRight =
+          right1.allSourceFileInfos.size + right2.allSourceFileInfos.size
+
+        if (hybridScanEnabled && numSourceFilesLeft != numSourceFilesRight) {
+          // For append-only dataset, the number of source files can indicate the amount of data
+          // that an index covers the input relation.
```

Comment:
Is it b/c of perf that we are not actually calculating the sum of actual file sizes to cover delete scenario? Should we just check that this is actually an append scenario?

Reply:
Yes it's because of the perf. And for delete, we could pick the index based on the creation time if not considering the follow corner cases:
1) create index A
2) delete most of the data
3) create a new index B
4) "undelete" of 2) which preserves the metadata of each files

In this case, we need to pick A rather than B. So I am thinking of introducing some "context" for each candidate index to reuse the diff file list calculated in getCandidateIndex (and verify the diff is append-only or not).
BTW why does join use sort instead of max/min?